{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Pat\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2837.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16820974757205074\"  style=\"position: relative; width: 640px; height: 480px\">\n        <p id=\"3dmolwarning_16820974757205074\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.0.1/3Dmol-min.js');\n}\n\nvar viewer_16820974757205074 = null;\nvar warn = document.getElementById(\"3dmolwarning_16820974757205074\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16820974757205074 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_16820974757205074\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:6LU7\", viewer_16820974757205074, {}, function() {\nviewer_16820974757205074.zoomTo();\n\tviewer_16820974757205074.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\nviewer_16820974757205074.render();\n})\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_16820974757205074\"  style=\"position: relative; width: 640px; height: 480px\">\n",
       "        <p id=\"3dmolwarning_16820974757205074\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
       "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    //this is to ignore the existence of requirejs amd\n",
       "    var savedexports, savedmodule;\n",
       "    if (typeof exports !== 'undefined') savedexports = exports;\n",
       "    else exports = {}\n",
       "    if (typeof module !== 'undefined') savedmodule = module;\n",
       "    else module = {}\n",
       "\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "        exports = savedexports;\n",
       "        module = savedmodule;\n",
       "        resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.0.1/3Dmol-min.js');\n",
       "}\n",
       "\n",
       "var viewer_16820974757205074 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_16820974757205074\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_16820974757205074 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_16820974757205074\"),{backgroundColor:\"white\"});\n",
       "$3Dmol.download(\"pdb:6LU7\", viewer_16820974757205074, {}, function() {\n",
       "viewer_16820974757205074.zoomTo();\n",
       "\tviewer_16820974757205074.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\n",
       "viewer_16820974757205074.render();\n",
       "})\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<py3Dmol.view at 0x1bb7dd51420>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Levenshtein import distance\n",
    "from Bio import Align, pairwise2, SeqIO\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import seq3, gc_fraction, MeltingTemp as melting_temp, GC123, GC_skew, xGC_skew, nt_search\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.PDB import PDBParser, MMCIFParser\n",
    "import py3Dmol\n",
    "from collections import Counter\n",
    "# import os\n",
    "\n",
    "# here = os.path.dirname(os.path.abspath(__file__))\n",
    "\n",
    "dna_seq = Seq('ATGATCTCGTAA')\n",
    "\n",
    "record = SeqRecord(dna_seq, id=\"test\", annotations={\"molecule_type\": \"DNA\"})\n",
    "\n",
    "# print(Counter(seq))\n",
    "\n",
    "dna_frequency = Counter(dna_seq)\n",
    "\n",
    "# plt.bar(dna_frequency.keys(), dna_frequency.values())\n",
    "\n",
    "\n",
    "# Transcribe dna to mRna and then to an Amino Acid\n",
    "dna_seq.complement()\n",
    "\n",
    "m_rna = dna_seq.transcribe()\n",
    "\n",
    "amino_acid = m_rna.translate()\n",
    "\n",
    "# Direct Translation:\n",
    "aa = dna_seq.translate()\n",
    "\n",
    "# transcribe back to DNA:\n",
    "m_rna.back_transcribe()\n",
    "\n",
    "# Get full Amino Acid name\n",
    "seq3(aa)\n",
    "\n",
    "# Dna Codontable\n",
    "CodonTable.unambiguous_dna_by_name['Standard']\n",
    "\n",
    "# Rna CodonTable\n",
    "CodonTable.unambiguous_rna_by_name['Standard']\n",
    "\n",
    "\n",
    "# GC Contents in DNA\n",
    "# higher GC content indicates a relatively higher melting temperature\n",
    "gc_fraction(dna_seq)\n",
    "\n",
    "\n",
    "melting_temp.Tm_Wallace(dna_seq)\n",
    "melting_temp.Tm_GC(dna_seq)\n",
    "\n",
    "def highest_gc(seq1, seq2):\n",
    "    gc1 = gc_fraction(seq1) * 100\n",
    "    gc2 = gc_fraction(seq2) * 100\n",
    "    at1 = 100 - gc1\n",
    "    at2 = 100 - gc2\n",
    "    melting1 = melting_temp.Tm_GC(seq1)\n",
    "    melting2 = melting_temp.Tm_GC(seq2)\n",
    "    result1 = \"Sequence: {} , GC: {} , AT: {} , Temp: {}\".format(seq1, gc1, at1, melting1)\n",
    "    result2 = \"Sequence: {} , GC: {} , AT: {} , Temp: {}\".format(seq2, gc2, at2, melting2)\n",
    "\n",
    "    if gc1 > gc2:\n",
    "        return result1\n",
    "    else:\n",
    "        return result2\n",
    "    \n",
    "# GC Skew\n",
    "# print(GC123(dna_seq))\n",
    "\n",
    "# print(GC_skew(dna_seq, 10))\n",
    "\n",
    "# Sequence Alighnment\n",
    "seq1 = Seq('ACTCGT')\n",
    "seq2 = Seq('ATTCG')\n",
    "\n",
    "# Global Alighnments\n",
    "alignments = pairwise2.align.globalxx(seq1,seq2)\n",
    "# for a in alignments:\n",
    "#     print(format_alignment(*a))\n",
    "\n",
    "# local Alighnments\n",
    "loc_alignments = pairwise2.align.localxx(seq1,seq2)\n",
    "# for a in loc_alignments:\n",
    "#     print(format_alignment(*a))\n",
    "\n",
    "\n",
    "test_alignment = pairwise2.align.globalms(seq1,seq2, 2, -1, -0.5, -0.1)\n",
    "# for a in test_alignment:\n",
    "    #  print(format_alignment(*a))\n",
    "\n",
    "seqA = Seq('AAGGCTT')\n",
    "seqB = Seq('AAGGC')\n",
    "seqC = Seq('AAGGCAT')\n",
    "\n",
    "alignerLocal = Align.PairwiseAligner()\n",
    "alignerLocal.mode = \"local\"\n",
    "\n",
    "AvB = alignerLocal.align(seqA, seqB)\n",
    "BvC = alignerLocal.align(seqB, seqC)\n",
    "AvC = alignerLocal.align(seqA, seqC)\n",
    "# print('AvB: ' ,AvB.score/len(seqB)*100)\n",
    "# print('BvC: ' ,BvC.score/len(seqB)*100)\n",
    "# print('AvC: ' ,AvC.score/len(seqC)*100)\n",
    "\n",
    "# Hamming Distance fxn\n",
    "\n",
    "seq1 = Seq('ACTAT')\n",
    "seq2 = Seq('ACTTA')\n",
    "seq3 = Seq('ACTT')\n",
    "def hamming_distance(lhs,rhs):\n",
    "    return( len([(x,y) for x,y in zip(lhs,rhs) if x!=y]))\n",
    "hamming_distance(seq1, seq2)\n",
    "\n",
    "# 0 if match\n",
    "hamming_distance(seq1, seq1)\n",
    "hamming_distance(seq1, seq3)\n",
    "\n",
    "# Levenshtein Distance\n",
    "distance(str(seq1), str(seq2))\n",
    "\n",
    "# print(\"Hamming Distance\", hamming_distance(seq1, seq3))\n",
    "# print(\"Levenshtein Distance\", distance(str(seq1), str(seq3)))\n",
    "\n",
    "\n",
    "# stack overflow code here https://stackoverflow.com/questions/40822400/how-to-create-a-dotplot-of-two-dna-sequence-in-python\n",
    "def delta(x,y):\n",
    "    return 0 if x == y else 1\n",
    "\n",
    "def M(seq1,seq2,i,j,k):\n",
    "    return sum(delta(x,y) for x,y in zip(seq1[i:i+k],seq2[j:j+k]))\n",
    "\n",
    "def makeMatrix(seq1,seq2,k):\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    return [[M(seq1,seq2,i,j,k) for j in range(m-k+1)] for i in range(n-k+1)]\n",
    "\n",
    "def plotMatrix(M,t, seq1, seq2, nonblank = chr(10000), blank = ' '):\n",
    "    print(' |' + seq2)\n",
    "    print('-'*(2 + len(seq2)))\n",
    "    for label,row in zip(seq1,M):\n",
    "        line = ''.join(nonblank if s < t else blank for s in row)\n",
    "        print(label + '|' + line)\n",
    "\n",
    "def dotplot(seq1,seq2,k = 1,t = 1):\n",
    "    M = makeMatrix(seq1,seq2,k)\n",
    "    plotMatrix(M, t, seq1,seq2) #experiment with character choice\n",
    "\n",
    "# Sequence Alignment with Dot Plot\n",
    "seq1 = Seq('ACTTAG')\n",
    "seq2 = Seq('AC')\n",
    "\n",
    "\n",
    "def dotplotx(seq1,seq2,k = 1,t = 1):\n",
    "    plt.imshow(np.array(makeMatrix(seq1,seq2,1)))\n",
    "    plt.xticks(np.arange(len(list(seq1))),list(seq1))\n",
    "    plt.yticks(np.arange(len(list(seq1))),list(seq1))\n",
    "    plt.show()\n",
    "\n",
    "dna1 = Seq('ATGATCTCGTAA')\n",
    "dna2 = Seq('ATTATGTCGTAA')\n",
    "\n",
    "# dotplotx(dna1, dna2)\n",
    "\n",
    "# Query Data with file formats \n",
    "# for record in SeqIO.parse(\"sequence.fasta\", \"fasta\"):\n",
    "#     print(record)\n",
    "\n",
    "# Reading FASTA\n",
    "# filename = os.path.join(here, 'sequence.fasta')\n",
    "dna_record = SeqIO.read('sequence.fasta', \"fasta\")\n",
    "dna_seq = dna_record.seq\n",
    "\n",
    "# Reading Genbank\n",
    "# filename = os.path.join(here, 'sequence.gb')\n",
    "gb_dna_record = SeqIO.parse('sequence.gb', \"genbank\")\n",
    "    \n",
    "\n",
    "# Working with 3D structures\n",
    "parser = PDBParser()\n",
    "# filename = os.path.join(here, '6lu7.pdb')\n",
    "structure = parser.get_structure('6LU7', '6lu7.pdb')\n",
    "model = structure[0]\n",
    "\n",
    "# structure => model => chain => residue => atom\n",
    "# check for chains\n",
    "# for chain in model:\n",
    "#     for atom in chain:\n",
    "#         print(atom)\n",
    "\n",
    "# Using py3DMol\n",
    "view1 = py3Dmol.view(query='pdb:6LU7')\n",
    "view1.setStyle({'cartoon': {'color':'spectrum'}})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16821536467569122\"  style=\"position: relative; width: 640px; height: 480px\">\n        <p id=\"3dmolwarning_16821536467569122\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    //this is to ignore the existence of requirejs amd\n    var savedexports, savedmodule;\n    if (typeof exports !== 'undefined') savedexports = exports;\n    else exports = {}\n    if (typeof module !== 'undefined') savedmodule = module;\n    else module = {}\n\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n        exports = savedexports;\n        module = savedmodule;\n        resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.0.1/3Dmol-min.js');\n}\n\nvar viewer_16821536467569122 = null;\nvar warn = document.getElementById(\"3dmolwarning_16821536467569122\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16821536467569122 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_16821536467569122\"),{backgroundColor:\"white\"});\n$3Dmol.download(\"pdb:4ZS6\", viewer_16821536467569122, {}, function() {\nviewer_16821536467569122.zoomTo();\n\tviewer_16821536467569122.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\nviewer_16821536467569122.render();\n})\n});\n</script>",
      "text/html": [
       "<div id=\"3dmolviewer_16821536467569122\"  style=\"position: relative; width: 640px; height: 480px\">\n",
       "        <p id=\"3dmolwarning_16821536467569122\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
       "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
       "        </div>\n",
       "<script>\n",
       "\n",
       "var loadScriptAsync = function(uri){\n",
       "  return new Promise((resolve, reject) => {\n",
       "    //this is to ignore the existence of requirejs amd\n",
       "    var savedexports, savedmodule;\n",
       "    if (typeof exports !== 'undefined') savedexports = exports;\n",
       "    else exports = {}\n",
       "    if (typeof module !== 'undefined') savedmodule = module;\n",
       "    else module = {}\n",
       "\n",
       "    var tag = document.createElement('script');\n",
       "    tag.src = uri;\n",
       "    tag.async = true;\n",
       "    tag.onload = () => {\n",
       "        exports = savedexports;\n",
       "        module = savedmodule;\n",
       "        resolve();\n",
       "    };\n",
       "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
       "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
       "});\n",
       "};\n",
       "\n",
       "if(typeof $3Dmolpromise === 'undefined') {\n",
       "$3Dmolpromise = null;\n",
       "  $3Dmolpromise = loadScriptAsync('https://cdnjs.cloudflare.com/ajax/libs/3Dmol/2.0.1/3Dmol-min.js');\n",
       "}\n",
       "\n",
       "var viewer_16821536467569122 = null;\n",
       "var warn = document.getElementById(\"3dmolwarning_16821536467569122\");\n",
       "if(warn) {\n",
       "    warn.parentNode.removeChild(warn);\n",
       "}\n",
       "$3Dmolpromise.then(function() {\n",
       "viewer_16821536467569122 = $3Dmol.createViewer(document.getElementById(\"3dmolviewer_16821536467569122\"),{backgroundColor:\"white\"});\n",
       "$3Dmol.download(\"pdb:4ZS6\", viewer_16821536467569122, {}, function() {\n",
       "viewer_16821536467569122.zoomTo();\n",
       "\tviewer_16821536467569122.setStyle({\"cartoon\": {\"color\": \"spectrum\"}});\n",
       "viewer_16821536467569122.render();\n",
       "})\n",
       "});\n",
       "</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<py3Dmol.view at 0x2b8172934c0>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import py3Dmol\n",
    "view2 = py3Dmol.view(query='pdb:4ZS6')\n",
    "view2.setStyle({'cartoon': {'color':'spectrum'}})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Pat\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\site-packages\\Bio\\PDB\\StructureBuilder.py:89: PDBConstructionWarning: WARNING: Chain A is discontinuous at line 2837.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d409ccd2eeb94d82bcf042350d4046e3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview as nv\n",
    "from Bio.PDB import PDBParser\n",
    "\n",
    "parser = PDBParser()\n",
    "structure = parser.get_structure('6LU7', '6lu7.pdb')\n",
    "\n",
    "view = nv.show_biopython(structure)\n",
    "view\n",
    "\n",
    "# render_image() to render the actual image as a file\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<?xml version=\"1.0\"?>\n",
      "<!DOCTYPE BlastOutput PUBLIC \"-//NCBI//NCBI BlastOutput/EN\" \"http://www.ncbi.nlm.nih.gov/dtd/NCBI_BlastOutput.dtd\">\n",
      "<BlastOutput>\n",
      "  <BlastOutput_program>blastn</BlastOutput_program>\n",
      "  <BlastOutput_version>BLASTN 2.14.0+</BlastOutput_version>\n",
      "  <BlastOutput_reference>Stephen F. Altschul, Thomas L. Madden, Alejandro A. Sch&amp;auml;ffer, Jinghui Zhang, Zheng Zhang, Webb Miller, and David J. Lipman (1997), &quot;Gapped BLAST and PSI-BLAST: a new generation of protein database search programs&quot;, Nucleic Acids Res. 25:3389-3402.</BlastOutput_reference>\n",
      "  <BlastOutput_db>nt</BlastOutput_db>\n",
      "  <BlastOutput_query-ID>Query_40723</BlastOutput_query-ID>\n",
      "  <BlastOutput_query-def>No definition line</BlastOutput_query-def>\n",
      "  <BlastOutput_query-len>29803</BlastOutput_query-len>\n",
      "  <BlastOutput_param>\n",
      "    <Parameters>\n",
      "      <Parameters_expect>10</Parameters_expect>\n",
      "      <Parameters_sc-match>2</Parameters_sc-match>\n",
      "      <Parameters_sc-mismatch>-3</Parameters_sc-mismatch>\n",
      "      <Parameters_gap-open>5</Parameters_gap-open>\n",
      "      <Parameters_gap-extend>2</Parameters_gap-extend>\n",
      "      <Parameters_filter>L;m;</Parameters_filter>\n",
      "    </Parameters>\n",
      "  </BlastOutput_param>\n",
      "<BlastOutput_iterations>\n",
      "<Iteration>\n",
      "  <Iteration_iter-num>1</Iteration_iter-num>\n",
      "  <Iteration_query-ID>Query_40723</Iteration_query-ID>\n",
      "  <Iteration_query-def>No definition line</Iteration_query-def>\n",
      "  <Iteration_query-len>29803</Iteration_query-len>\n",
      "<Iteration_hits>\n",
      "</Iteration_hits>\n",
      "  <Iteration_stat>\n",
      "    <Statistics>\n",
      "      <Statistics_db-num>92292742</Statistics_db-num>\n",
      "      <Statistics_db-len>1116908174805</Statistics_db-len>\n",
      "      <Statistics_hsp-len>0</Statistics_hsp-len>\n",
      "      <Statistics_eff-space>0</Statistics_eff-space>\n",
      "      <Statistics_kappa>0.41</Statistics_kappa>\n",
      "      <Statistics_lambda>0.625</Statistics_lambda>\n",
      "      <Statistics_entropy>0.78</Statistics_entropy>\n",
      "    </Statistics>\n",
      "  </Iteration_stat>\n",
      "  <Iteration_message>[blastsrv4.REAL]: Error: CPU usage limit was exceeded, resulting in SIGXCPU (24).</Iteration_message>\n",
      "</Iteration>\n",
      "</BlastOutput_iterations>\n",
      "</BlastOutput>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.Blast import NCBIWWW\n",
    "from Bio import SeqIO\n",
    "\n",
    "covid_record = SeqIO.read(\"covid_sequence_MT385497.fasta\",\"fasta\") \n",
    "covid_dna = covid_record.seq\n",
    "\n",
    "with NCBIWWW.qblast(\"blastn\", \"nt\", covid_dna) as res_handle:\n",
    "    with open(\"result_blast_covid.xml\", \"w\") as xml_file:\n",
    "        print(res_handle.getvalue())\n",
    "        xml_file.write(res_handle.getvalue())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
