{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from Levenshtein import distance\n",
    "from Bio import Align, pairwise2, SeqIO\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqUtils import seq3, gc_fraction, MeltingTemp as melting_temp, GC123, GC_skew, xGC_skew, nt_search\n",
    "from Bio.Data import CodonTable\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.PDB import PDBParser, MMCIFParser\n",
    "import py3Dmol\n",
    "from collections import Counter\n",
    "# import os\n",
    "\n",
    "# here = os.path.dirname(os.path.abspath(__file__))\n",
    "\n",
    "dna_seq = Seq('ATGATCTCGTAA')\n",
    "\n",
    "record = SeqRecord(dna_seq, id=\"test\", annotations={\"molecule_type\": \"DNA\"})\n",
    "\n",
    "# print(Counter(seq))\n",
    "\n",
    "dna_frequency = Counter(dna_seq)\n",
    "\n",
    "# plt.bar(dna_frequency.keys(), dna_frequency.values())\n",
    "\n",
    "\n",
    "# Transcribe dna to mRna and then to an Amino Acid\n",
    "dna_seq.complement()\n",
    "\n",
    "m_rna = dna_seq.transcribe()\n",
    "\n",
    "amino_acid = m_rna.translate()\n",
    "\n",
    "# Direct Translation:\n",
    "aa = dna_seq.translate()\n",
    "\n",
    "# transcribe back to DNA:\n",
    "m_rna.back_transcribe()\n",
    "\n",
    "# Get full Amino Acid name\n",
    "seq3(aa)\n",
    "\n",
    "# Dna Codontable\n",
    "CodonTable.unambiguous_dna_by_name['Standard']\n",
    "\n",
    "# Rna CodonTable\n",
    "CodonTable.unambiguous_rna_by_name['Standard']\n",
    "\n",
    "\n",
    "# GC Contents in DNA\n",
    "# higher GC content indicates a relatively higher melting temperature\n",
    "gc_fraction(dna_seq)\n",
    "\n",
    "\n",
    "melting_temp.Tm_Wallace(dna_seq)\n",
    "melting_temp.Tm_GC(dna_seq)\n",
    "\n",
    "def highest_gc(seq1, seq2):\n",
    "    gc1 = gc_fraction(seq1) * 100\n",
    "    gc2 = gc_fraction(seq2) * 100\n",
    "    at1 = 100 - gc1\n",
    "    at2 = 100 - gc2\n",
    "    melting1 = melting_temp.Tm_GC(seq1)\n",
    "    melting2 = melting_temp.Tm_GC(seq2)\n",
    "    result1 = \"Sequence: {} , GC: {} , AT: {} , Temp: {}\".format(seq1, gc1, at1, melting1)\n",
    "    result2 = \"Sequence: {} , GC: {} , AT: {} , Temp: {}\".format(seq2, gc2, at2, melting2)\n",
    "\n",
    "    if gc1 > gc2:\n",
    "        return result1\n",
    "    else:\n",
    "        return result2\n",
    "    \n",
    "# GC Skew\n",
    "# print(GC123(dna_seq))\n",
    "\n",
    "# print(GC_skew(dna_seq, 10))\n",
    "\n",
    "# Sequence Alighnment\n",
    "seq1 = Seq('ACTCGT')\n",
    "seq2 = Seq('ATTCG')\n",
    "\n",
    "# Global Alighnments\n",
    "alignments = pairwise2.align.globalxx(seq1,seq2)\n",
    "# for a in alignments:\n",
    "#     print(format_alignment(*a))\n",
    "\n",
    "# local Alighnments\n",
    "loc_alignments = pairwise2.align.localxx(seq1,seq2)\n",
    "# for a in loc_alignments:\n",
    "#     print(format_alignment(*a))\n",
    "\n",
    "\n",
    "test_alignment = pairwise2.align.globalms(seq1,seq2, 2, -1, -0.5, -0.1)\n",
    "# for a in test_alignment:\n",
    "    #  print(format_alignment(*a))\n",
    "\n",
    "seqA = Seq('AAGGCTT')\n",
    "seqB = Seq('AAGGC')\n",
    "seqC = Seq('AAGGCAT')\n",
    "\n",
    "alignerLocal = Align.PairwiseAligner()\n",
    "alignerLocal.mode = \"local\"\n",
    "\n",
    "AvB = alignerLocal.align(seqA, seqB)\n",
    "BvC = alignerLocal.align(seqB, seqC)\n",
    "AvC = alignerLocal.align(seqA, seqC)\n",
    "# print('AvB: ' ,AvB.score/len(seqB)*100)\n",
    "# print('BvC: ' ,BvC.score/len(seqB)*100)\n",
    "# print('AvC: ' ,AvC.score/len(seqC)*100)\n",
    "\n",
    "# Hamming Distance fxn\n",
    "\n",
    "seq1 = Seq('ACTAT')\n",
    "seq2 = Seq('ACTTA')\n",
    "seq3 = Seq('ACTT')\n",
    "def hamming_distance(lhs,rhs):\n",
    "    return( len([(x,y) for x,y in zip(lhs,rhs) if x!=y]))\n",
    "hamming_distance(seq1, seq2)\n",
    "\n",
    "# 0 if match\n",
    "hamming_distance(seq1, seq1)\n",
    "hamming_distance(seq1, seq3)\n",
    "\n",
    "# Levenshtein Distance\n",
    "distance(str(seq1), str(seq2))\n",
    "\n",
    "# print(\"Hamming Distance\", hamming_distance(seq1, seq3))\n",
    "# print(\"Levenshtein Distance\", distance(str(seq1), str(seq3)))\n",
    "\n",
    "\n",
    "# stack overflow code here https://stackoverflow.com/questions/40822400/how-to-create-a-dotplot-of-two-dna-sequence-in-python\n",
    "def delta(x,y):\n",
    "    return 0 if x == y else 1\n",
    "\n",
    "def M(seq1,seq2,i,j,k):\n",
    "    return sum(delta(x,y) for x,y in zip(seq1[i:i+k],seq2[j:j+k]))\n",
    "\n",
    "def makeMatrix(seq1,seq2,k):\n",
    "    n = len(seq1)\n",
    "    m = len(seq2)\n",
    "    return [[M(seq1,seq2,i,j,k) for j in range(m-k+1)] for i in range(n-k+1)]\n",
    "\n",
    "def plotMatrix(M,t, seq1, seq2, nonblank = chr(10000), blank = ' '):\n",
    "    print(' |' + seq2)\n",
    "    print('-'*(2 + len(seq2)))\n",
    "    for label,row in zip(seq1,M):\n",
    "        line = ''.join(nonblank if s < t else blank for s in row)\n",
    "        print(label + '|' + line)\n",
    "\n",
    "def dotplot(seq1,seq2,k = 1,t = 1):\n",
    "    M = makeMatrix(seq1,seq2,k)\n",
    "    plotMatrix(M, t, seq1,seq2) #experiment with character choice\n",
    "\n",
    "# Sequence Alignment with Dot Plot\n",
    "seq1 = Seq('ACTTAG')\n",
    "seq2 = Seq('AC')\n",
    "\n",
    "\n",
    "def dotplotx(seq1,seq2,k = 1,t = 1):\n",
    "    plt.imshow(np.array(makeMatrix(seq1,seq2,1)))\n",
    "    plt.xticks(np.arange(len(list(seq1))),list(seq1))\n",
    "    plt.yticks(np.arange(len(list(seq1))),list(seq1))\n",
    "    plt.show()\n",
    "\n",
    "dna1 = Seq('ATGATCTCGTAA')\n",
    "dna2 = Seq('ATTATGTCGTAA')\n",
    "\n",
    "# dotplotx(dna1, dna2)\n",
    "\n",
    "# Query Data with file formats \n",
    "# for record in SeqIO.parse(\"sequence.fasta\", \"fasta\"):\n",
    "#     print(record)\n",
    "\n",
    "# Reading FASTA\n",
    "# filename = os.path.join(here, 'sequence.fasta')\n",
    "dna_record = SeqIO.read('sequence.fasta', \"fasta\")\n",
    "dna_seq = dna_record.seq\n",
    "\n",
    "# Reading Genbank\n",
    "# filename = os.path.join(here, 'sequence.gb')\n",
    "gb_dna_record = SeqIO.parse('sequence.gb', \"genbank\")\n",
    "    \n",
    "\n",
    "# Working with 3D structures\n",
    "parser = PDBParser()\n",
    "# filename = os.path.join(here, '6lu7.pdb')\n",
    "structure = parser.get_structure('6LU7', '6lu7.pdb')\n",
    "model = structure[0]\n",
    "\n",
    "# structure => model => chain => residue => atom\n",
    "# check for chains\n",
    "# for chain in model:\n",
    "#     for atom in chain:\n",
    "#         print(atom)\n",
    "\n",
    "# Using py3DMol\n",
    "view1 = py3Dmol.view(query='pdb:6LU7')\n",
    "view1.setStyle({'cartoon': {'color':'spectrum'}})\n",
    "view1.addModel()\n",
    "view1.show()\n",
    "\n",
    "# Create a Py3Dmol view object\n",
    "view = py3Dmol.view()\n",
    "\n",
    "# Define a small molecule in SMILES format\n",
    "smiles = 'C[C@@H](O)[C@H](N)C(=O)O'\n",
    "\n",
    "# Add the molecule to the viewer\n",
    "view.addModel(smiles, 'smiles')\n",
    "\n",
    "# Set the style of the molecule\n",
    "view.setStyle({'stick':{}})\n",
    "\n",
    "# Display the viewer\n",
    "view.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
